### DISCUSSION TITLE
Insights on Trino’s pagination logic

### DISCUSSION THREAD

**Discussion by kuripart** [trinodb/trino#20045] [Q&A] [answered]
Hi,

Opening this discussion from the slack question asked in #dev: https://trinodb.slack.com/archives/CP1MUNEUX/p1701371089507779

We are trying to understand how trino’s pagination logic calculates the total number of pages and the rows per page to be returned on the execution of a single query generally (irrespective of the connector used), eg: `select * from abc;`?

In the [docs](https://trino.io/docs/current/develop/client-protocol.html), it says:
`A GET to the nextUri attribute returns the next batch of query results.`

Would appreciate if someone could explain and give some stats on this calculation, along with details on what properties or limits that influence it. From our understanding it varies on the table or file size, and is random. Thanks.

**Comment by martint** [comment, marked as answer]
There's no calculation ahead of time since execution is pipelined and results are produced and sent to the client while the query is still running. There is a target payload size for the each response in the client protocol, which is controlled by the `targetResultSize` query parameter. It defaults to 1MB.

See:
* https://github.com/trinodb/trino/blob/ae7849e5b81b0c0eff7e980222502bf392a9a7f3/core/trino-main/src/main/java/io/trino/server/protocol/ExecutingStatementResource.java#L162
* https://github.com/trinodb/trino/blob/ae7849e5b81b0c0eff7e980222502bf392a9a7f3/core/trino-main/src/main/java/io/trino/server/protocol/ExecutingStatementResource.java#L216-L221
